avro2py/avro_types.py

```diff
@@ -0,0 +1,200 @@
+"""Representation of avro schema types."""
+from contextlib import suppress
+from enum import Enum as _Enum, auto, unique
+from typing import NamedTuple, List, Union, Optional, Any, NewType, Dict
+
+
+def case_insensitive(cls):
+    """Add case-insensitive .get method to Enum."""
+    @classmethod
+    def get(cls_, value):
+        return cls_[value.upper().replace('-', '_')]
+    cls.get = get
+    return cls
+
+
+DefinedType = NewType('DefinedType', str)  # http://avro.apache.org/docs/1.10.0/spec.html#schemas
+
+
+@case_insensitive
+@unique
+class Primitives(_Enum):
+    """http://avro.apache.org/docs/1.10.0/spec.html#schema_primitive"""
+    NULL = auto()
+    BOOLEAN = auto()
+    INT = auto()
+    LONG = auto()
+    FLOAT = auto()
+    DOUBLE = auto()
+    BYTES = auto()
+    STRING = auto()
+
+
+@case_insensitive
+@unique
+class LogicalTypes(_Enum):
+    DECIMAL = auto()
+    DATE = auto()
+    TIME_MILLIS = auto()
+    TIME_MICROS = auto()
+    TIMESTAMP_MILLIS = auto()
+    TIMESTAMP_MICROS = auto()
+    DURATION = auto()
+    UUID = auto()
+
+
+class LogicalType(NamedTuple):
+    type: Primitives
+    logical_type: LogicalTypes
+    precision: Optional[int] = None
+    scale: Optional[int] = None
+
+
+AvroPrimitives = Union[Primitives, LogicalType]
+AvroType = Union[AvroPrimitives, NamedTuple, List, DefinedType]
+
+
+class Record(NamedTuple):
+    """http://avro.apache.org/docs/1.10.0/spec.html#Records"""
+    name: str
+    namespace: str
+    fields: List['Field']
+    doc: Optional[str] = None
+    aliases: List[str] = []
+
+
+class Field(NamedTuple):
+    name: str
+    type: AvroType
+
+
+class Enum(NamedTuple):
+    """http://avro.apache.org/docs/1.10.0/spec.html#Enums"""
+    name: str
+    namespace: str
+    symbols: List[str]
+    aliases: List[str] = []
+    doc: Optional[str] = None
+    default: Optional[str] = None
+
+
+class Array(NamedTuple):
+    """http://avro.apache.org/docs/1.10.0/spec.html#Arrays"""
+    items: AvroType
+    default: Optional[Any] = None
+
+
+class Map(NamedTuple):
+    """http://avro.apache.org/docs/1.10.0/spec.html#Maps"""
+    values: AvroType
+    default: Optional[Any] = None  # technically has to have the type specified by "items", but hard to capture
+
+
+class Fixed(NamedTuple):
+    """http://avro.apache.org/docs/1.10.0/spec.html#Fixed"""
+    name: str
+    namespace: str
+    size: int
+    aliases: List[str] = []
+
+
+def parse_into_types(schema: Union[Dict[str, Any], str, List], parent_namespace: Optional[str] = None) \
```

Comment:
i'm going to do an obnoxious thing, and just comment with "unit" on any function (like this one) that i think: 
1. wouldn't be terribly difficult to unit test
2. i would not feel confident changing without unit
3. i think has no unit currently

i don't expect them all to be addressed, just going to point them out. i think most of the code is probably exercised indirectly, so i don't think things are wrong. 

btw did you run coverage and just check that all the branches are hit?

Reply:
The coverage from the sample schema is 94% for the `codegen` module; the things that aren't getting exercised are the `fixed` type (of which I don't have an example) and some handling for edge-cases that I'm not sure are in-spec.